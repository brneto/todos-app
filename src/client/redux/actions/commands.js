import { createActions } from 'redux-actions';
import { metaFilterListCreator } from './utils';

// Pure Command actions
// Naming Convention: <imperative verb><subject>
export const { addTodoToList, removeTodoFromList } =
  createActions({
    ADD_TODO_TO_LIST: metaFilterListCreator,
    REMOVE_TODO_FROM_LIST: metaFilterListCreator,
  });
