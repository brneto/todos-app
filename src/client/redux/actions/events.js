import { createActions } from 'redux-actions';

// Event actions
// Naming Convention:
// - Instant event: <present-tense verb>
// - Progressive event start: <present-continuous verb>[<subject>]
// - Progressive event end: <past-tense verb>[<subject>]
export const {
  fetchingTodos,
  fetchedTodos,
} = createActions(
  'FETCHING_TODOS',
  'FETCHED_TODOS'
);
