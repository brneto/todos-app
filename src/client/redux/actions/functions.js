import { identity } from '../reducers/util';

// actionCreator :: type -> (data, filter | undefined) -> {
//   type: 'COMMAND' | 'DOCUMENT' | 'EVENT' | 'SIDE_EFFECT',
//   payload: data,
//   meta: { type: type | [, filter: filter] }
// }

export const createActionFunction = type => [
  identity,
  (payload, filter) => filter ? ({ type, filter }) : ({ type })
];
