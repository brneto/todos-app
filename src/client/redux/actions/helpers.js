
import { identity } from 'ramda';

// actionCreator :: type -> (data, filter | undefined) -> {
//   type: 'ACTION_TYPE',
//   payload: data,
//   meta: { type: type | [, filter: filter] }
// }
const
  createActionFunction = type =>
    [identity, (payload, filter) => filter ? ({ type, filter }) : ({ type })];

export { createActionFunction };
