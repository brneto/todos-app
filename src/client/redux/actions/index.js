import { createActions } from 'redux-actions';

export const {
  fetchTodos,
  addTodo,
  toggleTodo,

  setToggleFetching,

  setFetchTodos,
  setAddTodo,
  setToggleTodo,
} = createActions(
  {
    SET_TOGGLE_FETCHING: [
      () => null,
      (payload, meta) => ({ filter: payload instanceof Error ? meta : payload })
    ],
    SET_FETCH_TODOS: [
      payload => payload,
      (todos, filter) => ({ filter })
    ],
  },
  'SET_ADD_TODO',
  'SET_TOGGLE_TODO',
  'FETCH_TODOS',
  'ADD_TODO',
  'TOGGLE_TODO',
);
