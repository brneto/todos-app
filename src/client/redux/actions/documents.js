import { createActions } from 'redux-actions';
import { createActionFunction } from './helpers';

const messageType = 'DOCUMENT';

// Document actions
// Naming Convention: <subject><past-tense verb>
export const {
  todosFetched, todoAdded, todoToggled
} = createActions({
  TODOS_FETCHED: createActionFunction(messageType),
  TODO_ADDED: createActionFunction(messageType),
  TODO_TOGGLED: createActionFunction(messageType),
});
