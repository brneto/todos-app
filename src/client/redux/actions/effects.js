import { createActions } from 'redux-actions';
import { createActionFunction } from './helpers';

const messageType = 'SIDE_EFFECT';

// Side Effect Command actions
// Naming Convention: <imperative verb><subject>
export const {
  fetchTodos, addTodo, toggleTodo
} = createActions({
  'FETCH_TODOS': createActionFunction(messageType),
  'ADD_TODO': createActionFunction(messageType),
  'TOGGLE_TODO': createActionFunction(messageType),
});
