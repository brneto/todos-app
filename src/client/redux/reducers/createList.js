import { combineReducers } from 'redux';
import { combineActions, handleActions } from 'redux-actions';
import { prop, identity, equals, always } from 'ramda';
import { produce } from 'immer';
import { createSelector } from 'reselect';
import { events, documents } from '../actions';

const createList = filter => {
  const isFilter = equals(filter);

  const ids = handleActions(
      {
        [documents.todosFetched]: {
          next: produce((draft, { payload, meta }) =>
            isFilter(meta.filter) ? payload.result : [] //return an entirely new state
          ),
          throw: always([])
        },
        [documents.todoAdded]: {
          next: produce((draft, { payload, meta }) => {
            if (!isFilter('completed') && isFilter(meta.filter))
              draft.push(payload.result); //modify the current draft state
          }),
        },
      },
      [] // Initial state
    );

  const isFetching = handleActions(
      {
        [events.fetchingTodos]: produce((draft, { payload }) => {
          if (isFilter(payload)) return true; //return an entirely new state
        }),
        [events.fetchedTodos]: produce((draft, { payload }) => {
          if (isFilter(payload)) return false; //return an entirely new state
        }),
      },
      false // Initial state
    );

  const
    { todosFetched, todoAdded, todoToggled } = documents,
    error = handleActions(
      {
        [combineActions(todosFetched, todoAdded, todoToggled)]: {
          next: always(null),
          throw: produce((draft, { payload, meta }) =>
            isFilter(meta.filter) ? payload : null //return an entirely new state
          )
        },
      },
      null // Initial state
    );

  return combineReducers({
    ids,
    isFetching,
    error,
  });
};

// SELECTORS
// prop :: s -> {s: a} -> a | Undefined
const
  getIds = createSelector([prop('ids')], identity),
  getIsFetching = createSelector([prop('isFetching')], identity),
  getError = createSelector([prop('error')], identity);

export {
  createList as default,
  getIds,
  getIsFetching,
  getError,
};
