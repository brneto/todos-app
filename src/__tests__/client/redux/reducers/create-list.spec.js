import deepFreeze from 'deep-freeze';
import createList from '../../../../client/redux/reducers/create-list';
import { commands, events, documents } from '../../../../client/redux/actions';

describe('reducers/createList', () => {
  const
    filter = 'active',
    listByFilter = createList(filter),
    initialState = deepFreeze(listByFilter(undefined, { type: 'INIT' })),
    error = new Error('Boom!'),
    baseData = { result: [1, 2, 3] },
    baseState = deepFreeze(listByFilter(initialState, documents.todosFetched(baseData, filter)));

  it('should handle unknown actions', () => {
    // when
    const newState = listByFilter(initialState, { type: 'FAKE' });

    // then
    expect(newState).toBe(initialState);
  });

  it('should add fetched todos to empty list', () => {
    // when
    const newState = listByFilter(initialState, documents.todosFetched(baseData, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should add fetched todos to non-empty list', () => {
    // given
    const testData = { result: [4, 5, 6] };

    // when
    const newState = listByFilter(baseState, documents.todosFetched(testData, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should add fetched todos saves error message', () => {
    // when
    const newState = listByFilter(baseState, documents.todosFetched(error, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should add todo to empty list', () => {
    // given
    const testData = { result: 4 };

    // when
    const newState = listByFilter(initialState, documents.todoAdded(testData, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should add todo to non-empty list', () => {
    // given
    const testData = { result: 4 };

    // when
    const newState = listByFilter(baseState, documents.todoAdded(testData, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should add todo save error message', () => {
    // when
    const newState = listByFilter(baseState, documents.todoAdded(error, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should add a todo to the list', () => {
    // given
    const data = { result: 5 };

    // when
    const newState = listByFilter(baseState, commands.addTodoToList(data, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should remove a todo from the list', () => {
    // given
    const data = { result: 1 };

    // when
    const newState = listByFilter(baseState, commands.removeTodoFromList(data, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should toggle todo saves error message', () => {
    // given
    const
      filter = 'all',
      listByFilter = createList(filter),
      initialState = deepFreeze(listByFilter(undefined, { type: 'INIT' })),
      baseData = { result: [1, 2, 3] },
      baseState = deepFreeze(listByFilter(initialState, documents.todosFetched(baseData, filter)));

    // when
    const newState = listByFilter(baseState, documents.todoToggled(error, filter));

    // then
    expect(newState).toMatchSnapshot();
  });

  it('should set fetching state to true', () => {
    // when
    const newState = listByFilter(baseState, events.fetchingTodos(filter));

    // then
    expect(newState.isFetching).toBe(true);
  });

  it('should set fetching state to false', () => {
    // when
    const newState = listByFilter(baseState, events.fetchedTodos(filter));

    // then
    expect(newState.isFetching).toBe(false);
  });
});
