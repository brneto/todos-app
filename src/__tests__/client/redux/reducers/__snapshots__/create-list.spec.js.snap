// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/createList should add a todo to the list 1`] = `
Object {
  "error": null,
  "fetchStatus": "idle",
  "ids": Array [
    1,
    2,
    3,
    5,
  ],
}
`;

exports[`reducers/createList should add fetched todos saves error message 1`] = `
Object {
  "error": [Error: Boom!],
  "fetchStatus": "idle",
  "ids": Array [],
}
`;

exports[`reducers/createList should add fetched todos to empty list 1`] = `
Object {
  "error": null,
  "fetchStatus": "idle",
  "ids": Array [
    1,
    2,
    3,
  ],
}
`;

exports[`reducers/createList should add fetched todos to non-empty list 1`] = `
Object {
  "error": null,
  "fetchStatus": "idle",
  "ids": Array [
    4,
    5,
    6,
  ],
}
`;

exports[`reducers/createList should add todo save error message 1`] = `
Object {
  "error": [Error: Boom!],
  "fetchStatus": "idle",
  "ids": Array [
    1,
    2,
    3,
  ],
}
`;

exports[`reducers/createList should add todo to empty list 1`] = `
Object {
  "error": null,
  "fetchStatus": "idle",
  "ids": Array [
    4,
  ],
}
`;

exports[`reducers/createList should add todo to non-empty list 1`] = `
Object {
  "error": null,
  "fetchStatus": "idle",
  "ids": Array [
    1,
    2,
    3,
    4,
  ],
}
`;

exports[`reducers/createList should remove a todo from the list 1`] = `
Object {
  "error": null,
  "fetchStatus": "idle",
  "ids": Array [
    2,
    3,
  ],
}
`;

exports[`reducers/createList should toggle todo saves error message 1`] = `
Object {
  "error": [Error: Boom!],
  "fetchStatus": "idle",
  "ids": Array [
    1,
    2,
    3,
  ],
}
`;
